_posts/2021-02-01-Dynarec.md

```diff
@@ -0,0 +1,428 @@
+[Dolphin](https://www.alchemistowl.org/pocorgtfo/pocorgtfo06.pdf#page=11) and
+[Dynarmic](https://github.com/MerryMage/dynarmic/blob/master/docs/ReturnStackBufferOptimization.md)
+
+## Static Branch Linking
+
+So far, these optimizations have been built on assumptions of the emulator path,
+but information available for specific blocks at compile time can be used for
+further performance improvements. The most obvious and important of these are
+static branches, which almost always jump to the same one or two host addresses.
+These are usually so common, in fact, that JITs can replace the ending lookup
+for these blocks with a direct host branch, or no branch at all (fallthrough).
+This reduces the pressure on the host indirect branch predictor, but makes
+modification and invalidation more difficult than with a single global lookup
+table. Depending on how complicated your lookup process - for example, if it
+requires virtual address translation, checking pipeline states, probing a hash
+table, or other complex logic - it could also save memory pressure and
+computation.
+
+One way to implement static branch linking involves replacing the lookup part of
+the epilogue in blocks ending with static branches. The replacement epilogue
+initially includes a few nops followed by a short stub that passes the host
+address of the nops and the guest address of the next block to a linker
+function. The linker function can then lookup the host address of the next block
+using the passed guest address, then write a conditional branch to that host
+address at the host address of the nops, which should have been sized so that
+this write does not affect the linker call site after it. After the write the
+linker should make a note of the host address it just wrote to in a linker table
+so that the write can be undone when the target is invalidated. It can then jump
+to the host address of the next block to continue execution as usual. If the
+next block has not been compiled yet, then we may have to skip the writing and
+exit the JIT to compile, then wait for the next time the block is called to
+perform any linking.
+
+If the linked block ends in a guest conditional branch, the linker function may
+be called a second time from the same site, when the other arm of the branch is
+taken. Since there are only two arms, we can assume the linker function cannot
+be called a third time from the same site, so we can now write an unconditional
+branch to the next host address over that linker call site, assuming there is
+enough memory space there before the end of the block. The easiest way to check
+this condition is to check whether the host address passed to the linker points
+to nops, or if these have already been overwritten. The following pseudocode
+shows what the epilogue of a block ending in a guest conditional branch might
+look like at the 3 stages in its "life cycle":
+
+```
+// initial state of epilogue
+START:
+nops;
```

Comment:
Do you think it's worth mentioning checking the time slice/downcount here? I suppose it depends on what you're emulating, but otherwise an infinite loop in the guest code will also lead to an infinite loop on the host.